crates/rpc/src/handlers/block.rs

```diff
@@ -272,3 +279,46 @@ pub async fn get_block_from_id(
 
     Ok(HttpResponse::Ok().json(BeaconVersionedResponse::new(beacon_block)))
 }
+
+/// Called by `/beacon/heads` to get fork choice leaves.
+#[get("/beacon/heads")]
+pub async fn get_beacon_heads(db: Data<ReamDB>) -> Result<impl Responder, ApiError> {
+    let base = db
+        .justified_checkpoint_provider()
+        .get()
+        .map_err(|_| ApiError::InternalError)?;
+    let root = base.root;
+
+    let mut blocks = HashMap::with_hasher(RandomState::default());
+    let store = Store {
+        db: db.get_ref().clone(),
+    };
+
+    store.filter_block_tree(root, &mut blocks).map_err(|err| {
+        error!("Failed to filter block tree, error: {err:?}");
+        ApiError::InternalError
+    })?;
+
+    let mut leaves = Vec::new();
+    for (block_root, block) in &blocks {
+        let children = db
+            .parent_root_index_multimap_provider()
+            .get(*block_root)
+            .map_err(|err| {
+                error!("Failed to get children, error: {err:?}");
+                ApiError::InternalError
+            })?
+            .unwrap_or_default();
+
+        let is_leaf = children.iter().all(|child| !blocks.contains_key(child));
+        if is_leaf {
+            leaves.push(BeaconHeadResponse {
+                root: block.tree_hash_root(),
+                slot: block.slot,
+                execution_optimistic: false,
+            });
+        }
+    }
```

Comment:
```suggestion
    let mut leaves = vec![]j;
    let mut referenced_parents = HashSet::new();
    
    for block in blocks.values() {
        referenced_parents.insert(block.parent_root);
    }
    
    for (block_root, block) in &blocks {
        if !referenced_parents.contains(block_root) {
            leaves.push(BeaconHeadResponse {
                root: block.tree_hash_root(),
                slot: block.slot,
                execution_optimistic: false,
            });
        }
    }
```

I don't think the code here gets only the leaves. The code suggestion I have above should work ^